internal/urlclient/factory.go

```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+ * Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+// urlclient provides functions to integrate the client code in go-mod-core-contracts with application specific code
+package urlclient
+
+import (
+	"context"
+	"sync"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/urlclient/local"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/urlclient/retry"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/config"
+	"github.com/edgexfoundry/device-sdk-go/internal/endpoint"
+)
+
+// NewMetadata returns a URLClient that connects to a service attached to data.
+func NewData(isRegistry bool, wg *sync.WaitGroup, path string) interfaces.URLClient {
+	return newClient(isRegistry, wg, path, common.CurrentConfig.Clients[common.ClientData])
+}
+
+// NewMetadata returns a URLClient that connects to a service attached to metadata.
+func NewMetadata(isRegistry bool, wg *sync.WaitGroup, path string) interfaces.URLClient {
+	return newClient(isRegistry, wg, path, common.CurrentConfig.Clients[common.ClientMetadata])
+}
+
+// newClient is a factory function that uses pre-defined constants to reduce code duplication.
+func newClient(isRegistry bool, wg *sync.WaitGroup, path string, client common.ClientInfo) interfaces.URLClient {
+	if isRegistry {
+		return retry.New(
+			endpoint.New(
+				context.Background(),
+				wg,
+				&config.RegistryClient,
+				client.Name,
+				path,
+				15,
+			).Monitor(),
+			500,
+			10,
```

Comment:
The signature of retry.New is New(urlStream chan interfaces.URLStream, interval, timeout int)
https://github.com/edgexfoundry/go-mod-core-contracts/blob/5a23ac5557c84fd9545c1c160f7302282aa77822/clients/urlclient/retry/client.go#L35

Thus it should be (endpoint.Monitor(), 10, 500), not (endpoint.Monitor(), 500, 10)

Reply:
Also, it looks like we should put nanosecond instead of second as parameters
https://github.com/edgexfoundry/go-mod-core-contracts/blob/5a23ac5557c84fd9545c1c160f7302282aa77822/clients/urlclient/retry/client.go#L37-L38

https://golang.org/pkg/time/#Duration
A Duration represents the elapsed time between two instants as an int64 nanosecond count.

Reply:
The parameters are correct. The first integer is the rate at which the registry is polled for the interval in milliseconds, and the second integer is the number of seconds the algorithm retries for before throwing an error.

You're correct about the base unit of `Time.Duration()`, but in the actual algorithm we convert to millisecond and second, respectively.